modules/RTC/RTCUtils.js

```diff
@@ -1074,6 +1159,176 @@ class RTCUtils extends Listenable {
+
+                        // TODO(brian): Is this call needed? Why is this
+                        // happening at gUM time? Isn't there an event listener
+                        // for this?
+                        setAvailableDevices(umDevices, undefined);
+                        reject(new JitsiTrackError(
+                            error, constraints, umDevices));
+                    });
+            } catch (error) {
+                logger.error(`GUM failed: ${error}`);
+                reject(new JitsiTrackError(error, constraints, umDevices));
+            }
+        });
+    }
+
+    /**
+     * Acquire a display stream via the screenObtainer. This requires extra
+     * logic compared to use screenObtainer versus normal device capture logic
+     * in RTCUtils#_newGetUserMediaWithConstraints.
+     *
+     * @param {Object} desktopSharingExtensionExternalInstallation
+     * @param {string[]} desktopSharingSources
+     * @returns {Promise} A promise which will be resolved with an object whic
+     * contains the acquired display stream. If desktop sharing is not supported
+     * then a rejected promise will be returned.
+     */
+    _newGetDesktopMedia(
+            desktopSharingExtensionExternalInstallation,
+            desktopSharingSources) {
+        if (!screenObtainer.isSupported()) {
+            return Promise.reject(
+                new Error('Desktop sharing is not supported!'));
+        }
+
+        const desktopSharingOptions = {
+            ...desktopSharingExtensionExternalInstallation,
+            desktopSharingSources
+        };
+
+        return new Promise((resolve, reject) => {
+            screenObtainer.obtainStream(
+                desktopSharingOptions,
+                stream => {
+                    resolve(stream);
+                },
+                error => {
+                    reject(error);
+                });
+        });
+    }
+
+    /**
+     * Creates a constraints object to be passed into a call to getUserMedia.
+     *
+     * @param {Array} um - An array of user media types to get. The accepted
+     * types are "video", "audio", and "desktop."
+     * @param {Object} options - Various values to be added to the constraints.
+     * @param {string} options.cameraDeviceId - The device id for the video
+     * capture device to get video from.
+     * @param {Object} options.constraints - Default constraints object to use
+     * as a base for the returned constraints.
+     * @param {Object} options.desktopStream - The desktop source id from which
+     * to capture a desktop sharing video.
+     * @param {string} options.facingMode - Which direction the camera is
+     * pointing to.
+     * @param {string} options.micDeviceId - The device id for the audio capture
+     * device to get audio from.
+     * @private
+     * @returns {Object}
+     */
+    _newGetConstraints(um, options) {
+        const constraints = options.constraints
```

Comment:
i think adding a comment to call out the copy here would be good.  or (as discussed) a util function maybe?